src/registry/address/edit-validation.ts

```diff
@@ -0,0 +1,29 @@
+import {IntlShape} from 'react-intl';
+import {z} from 'zod';
+
+import {buildCommonSchema} from '@/registry/validation';
+
+import {POSTCODE_REGEX} from '../postcode/constants';
+
+// Constraints taken from the BRK API (apart from postcode which comes from our postcode component)
+const addressSchema = z.object({
+  postcode: z.string().regex(new RegExp(POSTCODE_REGEX)).optional(),
+  houseNumber: z
+    .string()
+    .regex(/^\d{1,5}$/)
+    .optional(),
+  houseLetter: z
+    .string()
+    .regex(/^[a-zA-Z]$/)
+    .optional(),
+  houseNumberAddition: z
+    .string()
+    .regex(/^([a-z,A-Z,0-9]){1,4}$/)
+    .optional(),
```

Comment:
Is something like `z.union([z.string().(...), z.literal('')])` more accurate?

Reply:
Not really, there's a nasty implementation detail that's relevant for the validation: when the input has no value (`''` in this case), formik will change that to `undefined` before passing it to validation. While this is counter-intuitive, the upstream issue/reason for this makes sense and it's something we have to live with. So the `z.literal('')` case will never actually happen, instead this ends up in the `.optional()` flow.

So what you currently have is the best reflection of the form inputs + validation. From a pure data modelling perspective, it could be better, but that's not what we are doing here.